src/IssueViz.UnitTests/Helpers/RunOnUIThreadTests.cs

```diff
@@ -0,0 +1,172 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.VisualStudio.Threading;
+using Moq;
+using SonarLint.VisualStudio.Integration.UnitTests;
+using SonarLint.VisualStudio.IssueVisualization.Helpers;
+
+namespace SonarLint.VisualStudio.IssueVisualization.UnitTests.Helpers
+{
+    [TestClass]
+    public class RunOnUIThreadTests
+    {
+        // These tests are slightly hacky in that they use "ThreadHelper.SetCurrentThreadAsUIThread()"
+        // to fool the VS thread helper into thinking the current thread is the UI thread.
+        // 
+        // However, these tests are quite limited as we can only test the cases where the
+        // caller is already on the main thread - we're not hacking/mocking/faking enough 
+        // of the VS setup for a call to "ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync()"
+        // to succeed.
+
+        [TestMethod]
+        public void Run_StartsOnUIThread_ExecuteOnUIThread()
+        {
+            ThreadHelper.SetCurrentThreadAsUIThread();
+
+            var initialThread = Thread.CurrentThread.ManagedThreadId;
+            int executionThreadId = -1;
+
+            // Act
+            RunOnUIThread.Run(() => executionThreadId = Thread.CurrentThread.ManagedThreadId);
+
+            executionThreadId.Should().Be(initialThread);
+            Thread.CurrentThread.ManagedThreadId.Should().Be(initialThread); // should still be on the "UI" thread
+        }
+
+        [TestMethod]
+        public async Task RunAsync_StartsOnUIThread_ExecuteOnUIThread()
+        {
+            ThreadHelper.SetCurrentThreadAsUIThread();
+
+            var initialThread = Thread.CurrentThread.ManagedThreadId;
+            int executionThreadId = -1;
+
+            // Act
+            await RunOnUIThread.RunAsync(() => executionThreadId = Thread.CurrentThread.ManagedThreadId);
+
+            executionThreadId.Should().Be(initialThread);
+            Thread.CurrentThread.ManagedThreadId.Should().Be(initialThread); // should still be on the "UI" thread
+        }
+    }
+
+    [TestClass]
+    public class RunOnUIThreadTests_Hacky
```

Comment:
Maybe "_VsMocking"?

Reply:
"Mocking" is a normal, safe thing to do in tests. I want it to be clear from the name that these tests are doing something non-standard and ugly. `RunUIThreadTests_HackVsInternals`?

Reply:
Yep sounds good